### ISSUE TITLE
Yolo11-cls heatmap generation

### ISSUE THREAD

**Issue by sasuke00** [ultralytics/ultralytics#23087]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Current heatmap generation does not able to show which part of the area affect the final decision. I would like to have the heatmap generated like the attached image which is more obviously.

<img width="844" height="224" alt="Image" src="https://github.com/user-attachments/assets/4d22a85c-ffca-4f11-89cc-35c45dbf3da5" />

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sasuke00, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for sharing the YOLO11-cls heatmap request and the reference image—super helpful 🧠🔥 This looks like a feature/visualization request around making classification heatmaps more interpretable (e.g., clearer “what regions influenced the decision”).

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

To help the team evaluate your request faster, please also share:
-Your exact command/code used for current heatmap generation (CLI or Python)
-Which `ultralytics` version you’re on, plus OS/GPU details
-Whether this is for classification only (YOLO11-cls) and what output style you want (single overlay, per-class maps, smoother blobs, etc.)
-A couple of example images + the current heatmap output vs. the desired output (like your screenshot) 📸

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response 🤖—an Ultralytics engineer will also assist soon once you share the details above.

**Comment by Y-T-G** [comment]
That appears to be EigenCAM or GradCAM

**Comment by sasuke00** [comment]
I had tried this，but it only take the last layer

**Comment by Y-T-G** [comment]
All visualizations require selecting a particular layer. You can select which layer you want to use in EigenCAM and GradCAM.

**Comment by sasuke00** [comment]
Ya, understand and I have tried it out, unfortunately there is none of the layer able to show the specific feature which affect the decision

**Comment by glenn-jocher** [comment]
If Grad-CAM/EigenCAM looks “non-specific” across *all* layers, that usually means the classifier is making a global/texture/background-based decision, so no layer will produce a sharp localization map until the model learns localized evidence (CAM can’t invent it). Try re-training with more object-centric crops (less background) and a larger `imgsz`, and when running CAM target the last *spatial* feature layer right before global pooling/classifier; to help pick the right layer, please share your CAM snippet + one example image + weights, and you can also sanity-check activations with the built-in [feature map visualization utility](https://docs.ultralytics.com/reference/utils/plotting/#ultralytics.utils.plotting.feature_visualization).

```python
from ultralytics import YOLO

m = YOLO("your_yolo11_cls.pt")
print(m.model)  # inspect modules and choose the last spatial layer as CAM target
```

**Comment by sasuke00** [comment]
`
  def apply_grad_cam(img_np, class_index):
      """
      Applies the Grad-CAM algorithm to generate and overlay a heatmap.
      
      Args:
          img_np (np.array): The original image (H x W x C, BGR format).
          class_index (int): The index of the predicted class.
          
      Returns:
          np.array: The original image with the heatmap overlaid, or None if data is missing.
      """
      global feature_map, gradients
      
      if feature_map is None or gradients is None:
          return None 
  
      # 1. Calculate weights (alpha_k): Global Average Pooling of gradients
      # Mean across H and W gives (1, Channels, 1, 1)
      weights = gradients.mean(dim=[2, 3], keepdim=True)
  
      # 2. Multiply weights with feature map and perform ReLU
      # Sum across the Channel dimension to get the CAM (1, 1, H, W)
      cam = (feature_map * weights).sum(dim=1, keepdim=True)
      cam = torch.relu(cam)
  
      # Convert CAM tensor to numpy array
      cam = cam.squeeze().cpu().numpy()
      
      # 3. Normalize the CAM
      cam = cam - cam.min()
      # Handle division by zero if cam is uniformly zero
      if cam.max() > 0:
          cam = cam / cam.max()
      else:
          return None 
  
      # 4. Resize CAM to the original image dimensions
      h, w = img_np.shape[:2]
      heatmap = cv2.resize(cam, (w, h))
  
      # 5. Convert heatmap to 3-channel BGR image
      heatmap = np.uint8(255 * heatmap)
      # Use cv2.COLORMAP_VIRIDIS for a good colormap
      heatmap = cv2.applyColorMap(heatmap, cv2.COLORMAP_VIRIDIS)
  
      # 6. Overlay the heatmap on the original image (70% original, 30% heatmap)
      superimposed_img = cv2.addWeighted(img_np, 0.7, heatmap, 0.3, 0)
      
      return superimposed_img
`


This is my current code by using Grad-CAM

**Comment by glenn-jocher** [comment]
Your Grad-CAM math is fine, but for Ultralytics `YOLO` wrappers the most common gotcha is that `YOLO(...)(img)` / `predict()` runs under `torch.no_grad()`, so you’ll only ever see “flat” or missing gradients unless you forward the underlying `m.model` with grads enabled; also make sure your `feature_map` is truly spatial (`B,C,H,W` with `H,W>1`), otherwise the CAM will be uniform by construction (this is a core limitation of CAM-style XAI on image-level classifiers, see [Ultralytics XAI overview](https://www.ultralytics.com/blog/all-you-need-to-know-about-explainable-ai)).  

```python
import torch
import torch.nn as nn
from ultralytics import YOLO

y = YOLO("your_yolo11_cls.pt")
net = y.model  # IMPORTANT: use the raw torch.nn.Module, not y.predict()/y(...)
net.eval()

# pick a *spatial* conv layer (last Conv2d that still has H,W > 1)
target = None
for name, m in net.named_modules():
    if isinstance(m, nn.Conv2d):
        target = (name, m)
print("Hooking:", target[0])

cache = {}
def fwd(_, __, out): cache["feat"] = out
def bwd(_, __, grad_out): cache["grad"] = grad_out[0]

h1 = target[1].register_forward_hook(fwd)
h2 = target[1].register_full_backward_hook(bwd)

# forward + backward on the class logit (pre-softmax)
x = ...  # torch tensor BCHW float on same device as net
logits = net(x)                 # expect shape [B, nc] for classify
score = logits[:, class_index].sum()
net.zero_grad(set_to_none=True)
score.backward()

print(cache["feat"].shape, cache["grad"].shape)  # should be (B,C,H,W) with H,W > 1
h1.remove(); h2.remove()
```

**Comment by sasuke00** [comment]
what if I am using the C2PSA layer to get the feature, can?

```
    model = YOLO(WEIGHTS_PATH)
    for param in model.model.parameters():
        param.requires_grad = True
    
    # --- 1. Find the target layer for Grad-CAM ---
    try:
        # Use model.model.model to access the Sequential list of layers
        # Index 9 is confirmed as the last C2PSA/backbone block for model variant
        target_layer = model.model.model[-2] 
        
    except IndexError:
        print("Warning: Target layer index 9 not found, trying last index.")
        target_layer = model.model.model[-1]
    except AttributeError:
        # Handles the initial TypeError
        print("Error: Could not find model.model.model structure. Grad-CAM cannot be initialized.")
        exit()
        
    print(f"Selected Grad-CAM target layer: {target_layer.__class__.__name__}") # Prints the type of module
    
    # Register the hooks
    h1 = target_layer.register_forward_hook(save_feature_map)
    h2 = target_layer.register_full_backward_hook(save_gradients)

```

**Comment by glenn-jocher** [comment]
Yes—`C2PSA` can be a Grad-CAM target as long as the tensor coming *out of that module* is truly spatial (`B,C,H,W` with `H,W>1`); if it’s already pooled down to `1x1` anywhere after it, the CAM will be uniform by construction (see the `C2PSA` definition in the [C2PSA module reference](https://docs.ultralytics.com/reference/nn/modules/block/#class-ultralyticsnnmodulesblockc2psa)). In practice you’ll usually get cleaner maps by hooking the **last `nn.Conv2d` inside** your chosen `C2PSA` block rather than the container module:

```python
import torch.nn as nn
from ultralytics import YOLO

y = YOLO(WEIGHTS_PATH)
net = y.model

block = net.model[-2]  # only if this is actually your C2PSA block

last_conv2d = None
for m in block.modules():
    if isinstance(m, nn.Conv2d):
        last_conv2d = m

assert last_conv2d is not None, "No nn.Conv2d found inside the selected block"

h1 = last_conv2d.register_forward_hook(save_feature_map)
h2 = last_conv2d.register_full_backward_hook(save_gradients)
```

**Comment by sasuke00** [comment]
Thanks

**Comment by glenn-jocher** [comment]
You’re welcome—if anything’s still not behaving as expected (e.g., hooks firing but `feat/grad` shapes aren’t `B,C,H,W`), please open a new issue (since this one is closed) with `yolo checks`, your weights, one sample image, and the exact `target_layer` name/index you’re hooking so we can reproduce it.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #25548](https://github.com/ultralytics/ultralytics/pull/25548), titled **“Replace predict feature-map dumps with class activation heatmaps.”**

Key changes include:

- Replaced raw layer-by-layer feature-map dumps with practical PyTorch class activation heatmaps using LayerCAM.
- Updated `visualize=True` during prediction to generate heatmap overlays alongside prediction outputs.
- Applied confidence and class filters, with support for visualizing a configurable number of top predictions.
- Limited activation-map generation to supported PyTorch workflows, excluding tensor sources, augmentation, and embedding workflows.
- Updated inference and documentation to reflect the new heatmap behavior.

These changes should provide more interpretable visual explanations by highlighting image regions that influence predictions, while ensuring visualizations focus on relevant detections.

The changes were implemented by [@Y-T-G](https://github.com/Y-T-G), with contributions from [@glenn-jocher](https://github.com/glenn-jocher).

To test the latest changes, install directly from the `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, you can wait for the next official PyPI release.

Please test this update and share whether it resolves the issue. Thank you 🙏 for reporting it, and any further feedback is welcome if the problem persists.

**Comment by glenn-jocher** [comment]
@sasuke00, @Y-T-G addressed this classification explainability request in merged PR #25548. YOLO classification prediction with `visualize=True` now saves a LayerCAM overlay showing which image regions contributed positively to the predicted class score.
